src/MassTransit/Transports/ConsumerAgent.cs

```diff
@@ -203,27 +203,52 @@ protected virtual async Task ActiveAndActualAgentsCompleted(StopContext context)
             }
         }
 
-        protected virtual bool IsDuplicate(TKey key)
+        protected virtual bool IsTrackable(TKey key)
         {
             return true;
         }
 
-        protected async Task Dispatch<TContext>(TKey key, TContext context, ReceiveLockContext receiveLock = default)
+        protected Task Dispatch<TContext>(TKey key, TContext context, ReceiveLockContext receiveLock)
             where TContext : BaseReceiveContext
         {
-            var added = _pending.TryAdd(key, context);
-            if (!added && IsDuplicate(key))
-                LogContext.Warning?.Log("Duplicate dispatch key {Key}", key);
+            var added = false;
+            var lockContext = receiveLock;
 
-            try
+            if (IsTrackable(key))
             {
-                await _dispatcher.Dispatch(context, receiveLock).ConfigureAwait(false);
+                lockContext = _pending.AddOrUpdate(key, _ =>
+                {
+                    added = true;
+                    context.Pending(receiveLock);
+                    return context;
+                }, (_, current) =>
+                {
+                    added = false;
+                    current.Pending(receiveLock);
+                    return current;
+                });
+
+                if (!added)
+                {
+                    LogContext.Warning?.Log("Duplicate dispatch key {Key}", key);
+                    return Task.CompletedTask;
+                }
             }
-            finally
+
+            async Task DispatchAsync()
             {
-                if (added)
-                    _pending.TryRemove(key, out _);
+                try
```

Comment:
Random thought, I wonder if it would be okay to use .ContinueWith() to remove the pending item and return the dispatcher Task directly (not the remove task), or if that would introduce some weird scheduling? I think if RunSynchronously is specified, it would run it right away. Avoids the async state machine creation.